File: algorithm_introduction/ss20_van_emde_boas_trees/cpp/van_emde_boas_tree.cpp
#include <math.h>
#include <memory>
#include <stdio.h>
#include <vector>

using namespace std;

struct vEBTree {
  int u = -1;
  int min = -1;
  int max = -1;
  int sqrt_u_upper = -1;
  int sqrt_u_lower = -1;
  shared_ptr<vEB_Tree> summary = nullptr;
  unique_ptr<shared_ptr<vEB_Tree>[]> cluster = nullptr;
  vEB_Tree(int u) : u(u) {
    sqrt_u_upper = int(ceil(sqrt(u)));
    sqrt_u_lower = int(sqrt(u));
    if (u == 2)
      return;
    summary = make_shared<vEB_Tree>(sqrt_u_upper);
    cluster = unique_ptr<shared_ptr<vEB_Tree>[]>(new shared_ptr<vEB_Tree>[sqrt_u_upper]);
    for (int i = 0; i < sqrt_u_upper; ++i) {
      cluster[i] = make_shared<vEB_Tree>(sqrt_u_lower);
    }
  }

  int high(int x) { return x / sqrt_u_lower; }

  int low(int x) { return x % sqrt_u_lower; }

  int index(int x, int y) { return x * sqrt_u_lower + y; }

  int vEB_tree_minimum() { return min; }

  int vEB_tree_maximum() { return max; }

  bool vEB_tree_member(int x) {
    if (x == min || x == max)
      return true;
    else if (u == 2)
      false;
    else
      return cluster[high(x)]->vEB_tree_member(low(x));
  }

  int vEB_tree_successor(int x) {
    if (u == 2) {
      if (x == 0 && max == 1)
        return 1;
      else
        return -1;
    } else if (min != -1 && x < min) {
      return min;
    } else {
      int max_low = cluster[high(x)]->vEB_tree_maximum();
      if (max_low != -1 && low(x) < max_low) {
        int offset = cluster[high(x)]->vEB_tree_successor(low(x));
        return index(high(x), offset);
      } else {
        int succ_cluster = summary->vEB_tree_successor(high(x));
        if (succ_cluster == -1)
          return -1;
        else {
          int offset = cluster[succ_cluster]->vEB_tree_minimum();
          return index(succ_cluster, offset);
        }
      }
    }
  }

  int vEB_tree_predecessor(int x) {
    if (u == 2) {
      if (x == 1 && min == 0) {
        return 0;
      } else {
        return -1;
      }
    } else if (max != -1 && x > max) {
      return max;
    } else {
      int min_low = cluster[high(x)]->vEB_tree_minimum();
      if (min_low != -1 && low(x) > min_low) {
        int offset = cluster[high(x)]->vEB_tree_predecessor(low(x));
        return index(high(x), offset);
      } else {
        int pred_cluster = summary->vEB_tree_predecessor(high(x));
        if (pred_cluster == -1) {
          if (min != -1 && x > min) {
            return min;
          } else {
            return -1;
          }
        } else {
          int offset = cluster[pred_cluster]->vEB_tree_maximum();
          return index(pred_cluster, offset);
        }
      }
    }
  }

  void vEB_empty_tree_insert(int x) {
    min = x;
    max = x;
  }

  void vEB_tree_insert(int x) {
    if (min == -1) {
      vEB_empty_tree_insert(x);
    } else {
      if (x < min) {
        swap(x, min);
      }
      if (u > 2) {
        if (cluster[high(x)]->vEB_tree_minimum() == -1) {
          summary->vEB_tree_insert(high(x));
          cluster[high(x)]->vEB_empty_tree_insert(low(x));
        } else {
          cluster[high(x)]->vEB_tree_insert(low(x));
        }
      }
      if (x > max) {
        max = x;
      }
    }
  }

  void vEB_tree_delete(int x) {
    if (min == max) {
      min = -1;
      max = -1;
    } else if (u == 2) {
      if (x == 0) {
        min = 1;
      } else {
        min = 0;
      }
      max = min;
    } else {
      if (x == min) {
        int first_cluster = summary->vEB_tree_minimum();
        x = index(first_cluster, cluster[first_cluster]->vEB_tree_minimum());
        min = x;
      }
      cluster[high(x)]->vEB_tree_delete(low(x));
      if (cluster[high(x)]->vEB_tree_minimum() == -1) {
        summary->vEB_tree_delete(high(x));
        if (x == max) {
          int summary_max = summary->vEB_tree_maximum();
          if (summary_max == -1) {
            max = min;
          } else {
            max = index(summary_max, cluster[summary_max]->vEB_tree_maximum());
          }
        }
      } else if (x == max) {
        max = index(high(x), cluster[high(x)]->vEB_tree_maximum());
      }
    }
  }
};

int main() {}

File: algorithm_introduction/ss20_van_emde_boas_trees/python/van_emde_boas_tree.py
import math


class vEBTree:
    def __init__(self, u: int):
        self.u = u
        self.min_val = -1
        self.max_val = -1
        self.sqrt_u_upper = math.ceil(math.sqrt(u))
        self.sqrt_u_lower = int(math.sqrt(u))
        self.summary = None
        self.cluster = None

        if self.u >= 2:
            self.summary = vEBTree(self.sqrt_u_upper)
            self.cluster = [vEBTree(self.sqrt_u_lower) for _ in range(self.sqrt_u_upper)]

    def high(self, x: int):
        return x / self.sqrt_u_lower

    def low(self, x: int):
        return x % self.sqrt_u_lower

    def index(self, x: int, y: int):
        return x * self.sqrt_u_lower + y

    def vEB_tree_minimum(self):
        return self.min_val

    def vEB_tree_maximum(self):
        return self.max_val

    def vEB_tree_member(self, x: int) -> bool:
        if x == self.min_val or x == self.max_val:
            return True
        elif self.u == 2:
            return False
        else:
            return self.cluster[self.high(x)].vEB_tree_member(self.low(x))

    def vEB_tree_successor(self, x: int) -> int:
        if self.u == 2:
            if x == 0 and self.max_val == 1:
                return 1
            else:
                return -1
        elif self.min_val != -1 and x < self.min_val:
            return self.min_val
        else:
            max_low = self.cluster[self.high(x)].vEB_tree_maximum()
            if max_low != -1 and self.low(x) < max_low:
                offset = self.cluster[self.high(x)].vEB_tree_successor(self.low(x))
                return self.index(self.high(x), offset)
            else:
                succ_cluster = self.summary.vEB_tree_successor(self.high(x))
                if succ_cluster == -1:
                    return -1
                else:
                    offset = self.cluster[succ_cluster].vEB_tree_minimum()
                    return self.index(succ_cluster, offset)

    def vEB_tree_predecessor(self, x: int) -> int:
        if self.u == 2:
            if x == 1 and self.min_val == 0:
                return 0
            else:
                return -1
        elif self.max_val != -1 and x > self.max_val:
            return self.max_val
        else:
            min_low = self.cluster[self.high(x)].vEB_tree_minimum()
            if min_low != -1 and x > min_low:
                offset = self.cluster[self.high(x)].vEB_tree_predecessor(self.low(x))
                return self.index(self.high(x), offset)
            else:
                pred_cluster = self.summary.vEB_tree_predecessor(self.high(x))
                if (pred_cluster == -1):
                    if self.min_val != -1 and self.min_val < x:
                        return self.min_val
                    else:
                        return -1
                else:
                    offset = self.cluster[pred_cluster].vEB_tree_maximum()
                    return self.index(pred_cluster, offset)

    def vEB_empty_tree_insert(self, x: int):
        self.min = x
        self.max = x

    def vEB_tree_insert(self, x: int):
        if self.min_val == -1:
            self.vEB_empty_tree_insert(x)
        else:
            if x < self.min_val:
                x, self.min_val = self.min_val, x
            if self.u > 2:
                if self.cluster[self.high(x)].vEB_tree_minimum() == -1:
                    self.summary.vEB_tree_insert(self.high(x))
                    self.cluster[self.high(x)].vEB_empty_tree_insert(self.low(x))
                else:
                    self.cluster[self.high(x)].vEB_tree_insert(self.low(x))
            if x > self.max_val:
                self.max_val = x

    def vEB_tree_delete(self, x: int):
        if self.min_val == self.max_val:
            self.min_val = -1
            self.max_val = -1
        elif self.u == 2:
            if x == 0:
                self.min_val = 1
            else:
                self.min_val = 0
        else:
            if x == self.min_val:
                first_cluster = self.summary.vEB_tree_minimum()
                x = self.index(first_cluster, self.cluster[first_cluster].vEB_tree_minimum())
                self.min_val = x
            self.cluster[self.high(x)].vEB_tree_delete(self.low(x))
            if self.clsuter[self.high(x)].vEB_tree_minimum() == -1:
                self.summary.vEB_tree_delete(self.high(x))
                if x == self.max_val:
                    summary_max = self.summary.vEB_tree_maximum()
                    if summary_max == -1:
                        self.max_val = self.min_val
                    else:
                        self.max_val = self.index(summary_max, self.cluster[summary_max].vEB_tree_maximum())
            elif x == self.max_val:
                self.max_val = self.index(self.high(x), self.cluster[self.high(x)].vEB_tree_maximum())
